### PR TITLE
Bulk fastt rich editors

### PR DESCRIPTION
**Pull request by danyx23**
This PR adds two kinds of special editors in the editor tab:
* A larger text editor for all text fields (especially useful for editing subtitles and other long texts)
* A colorScale editing control

The editing model is such that when you make a change in a sidebar editor you start seeing a confirm/cancel button where the first writes the changes back to the underlying config and the other one aborts the edit.

For color scales this change meant that it is no longer the individual fields inside the colorScale object that are shown as columns but instead the entire colorScale is shown as one column (or two, one for map and one for the chart). The design of these two buttons is not ideal and I'd be happy for better suggestions on where to place them.

Unfortunately copy/paste or drag to extend doesn't seem to work properly with colorScales (especially for the map colorScale). I  don't know why the grid is unhappy about these cells as it seems that not even the validate event gets fired. If you don't have objections then I would tackle this bug later.

I have not included list editors yet, even though they would be nice. includedEntities and excludedEntities would be quite simple to add but the more interesting one is selections. The problem with selections is that they span 3 fields that are all arrays that should be kept in sync. It would be quite a bit easier to do this once the have been unified to a single array of objects with 3 fields.

### CODE REVIEW

File: adminSiteClient/EditorColorScaleSection.tsx [hunk 1]
```diff
@@ -60,13 +65,16 @@ export class EditorColorScaleSection extends React.Component<{
 class ColorLegendSection extends React.Component<{
     scale: ColorScale
     features: EditorColorScaleSectionFeatures
+    onChange?: () => void
 }> {
     @action.bound onEqualSizeBins(isEqual: boolean) {
         this.props.scale.config.equalSizeBins = isEqual ? true : undefined
+        this.props.onChange ? this.props.onChange() : noop()
```
Comment (danielgavrilov) [inline adminSiteClient/EditorColorScaleSection.tsx:72]: This might be nicer:

```suggestion
        this.props.onChange?.()
```

**Comment by danielgavrilov** [comment]
Did you look into using a [mobx observer util](https://mobx.js.org/intercept-and-observe.html) for triggering `onChange`, instead of passing it as a parameter? A parameter feels fragile because we have to remember to invoke it every time, it's not in any way tied to the action of updating a value.

File: clientUtils/patchHelper.ts [hunk 2]
```diff
@@ -1,7 +1,25 @@
...
+    if (pointer.length === 0) throw new Error("Pointer must not be empty")
+    const key: string = pointer.shift()!
+    if (pointer.length === 0) {
+        json[key] = newValue
+        return json
+    }
+    if (!json[key]) {
+        json[key] = {}
+    }
+    return setValueRecursiveInplace(json[key], pointer, newValue)
```
Comment (danielgavrilov) [inline clientUtils/patchHelper.ts:20]: A bit worried about these edge cases: 
- if it's non-empty and a falsy value (`""`, `0`, `false`) it gets overwritten with `{}`
- if it's non-empty and a truthy value (`"abc"`, `1`), then we attempt to inject a property (even when it's a primitive, which becomes a noop)

It seems very unlikely they show up, but if they do, maybe better to raise an error? I think the only case we should set `{}` is when `json[key] === undefined`?

**Comment by danielgavrilov** [comment]
Nice to have: could the preview switch between map & chart tab as I edit the two different color scales?

https://user-images.githubusercontent.com/1308115/157029001-3adcf411-953c-4574-b5b2-9e17708ada28.mp4

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 3]
```diff
@@ -1162,7 +1465,13 @@ export class GrapherConfigGridEditor extends React.Component<GrapherConfigGridEd
         const { grapherElement } = this
         return (
             <div className="preview">
-                <h3>Interactive grapher preview</h3>
+                <h5>Interactive grapher preview</h5>
+                <Toggle
+                    label="Keep entity selection"
```
Comment (danielgavrilov) [inline adminSiteClient/GrapherConfigGridEditor.tsx:1470]: Could this label be explained a bit? The `title` below is helpful, but not visible in the UI

File: adminSiteClient/GrapherConfigGridEditorTypesAndUtils.tsx [hunk 4]
```diff
@@ -167,6 +173,7 @@ export interface ColumnDataSourceUnknown {
 }
 export type ColumnDataSource =
     | ColumnDataSourceFieldDescription
+    // | ColumnDataSourceMultipleFieldDescriptions
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditorTypesAndUtils.tsx:176]: I'm not sure if this should really be commented out?

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 5]
```diff
@@ -7,7 +7,7 @@ import {
     DropResult,
 } from "react-beautiful-dnd"
 import { Disposer, observer } from "mobx-react"
-import { observable, computed, action, runInAction, autorun } from "mobx"
+import { observable, computed, action, runInAction, autorun, toJS } from "mobx"
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditor.tsx:10]: nit: there are some unused imports, like this one

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 6]
```diff
@@ -91,6 +94,57 @@ import {
...
+import { UnControlled as CodeMirror } from "react-codemirror2"
+import jsonpointer from "json8-pointer"
+import { EditorColorScaleSection } from "./EditorColorScaleSection.js"
+import { MapChart } from "../grapher/mapCharts/MapChart.js"
+
+function HotColorScaleRenderer(props: Record<string, unknown>) {
+    return <div>Color scale</div>
+}
+
+class HotColorScaleEditor extends BaseEditorComponent<Record<string, never>> {
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditor.tsx:108]: you never really use this component.
I guess it was an experiment to show the color editor inline, which didn't work out?
should probably get rid of it then?

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 7]
```diff
@@ -244,27 +301,193 @@ export class GrapherConfigGridEditor extends React.Component<GrapherConfigGridEd
...
+        if (
+            selectedColumn === undefined ||
+            selectedColumn >= columnDataSources.length
+        ) {
+            return undefined
+        }
+        return columnDataSources[selectedColumn]
+    }
+
+    @computed private get currentColumnFieldDesription():
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditor.tsx:343]: typo: `description` hits again!

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 8]
```diff
@@ -244,27 +301,193 @@ export class GrapherConfigGridEditor extends React.Component<GrapherConfigGridEd
...
+            // Here we are setting the target column field directly at the grapher so the
+            // preview is updated. When the user clicks the save button we'll then check
+            // the value on grapher vs the value on the richDataRow and perform a proper
+            // update using doAction like we do on the grid editor commits.
+            setValueRecursiveInplace(grapher, pointer, value)
+        }
+    }
+
+    @action.bound
+    commitRichEditorChanges(performCommit: boolean) {
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditor.tsx:384]: I was confused what `performCommit` does (in a commit method), and was surprised that `performCommit = false` is a rollback.
You should clarify that a bit, or at least add a comment.

File: adminSiteClient/EditorColorScaleSection.tsx (hunk 1 above)
Comment (danyx23) [reply adminSiteClient/EditorColorScaleSection.tsx:72]: Oh neat, I didn't know that the null coalescing op could be used that way - thanks!

File: adminSiteClient/GrapherConfigGridEditor.tsx (hunk 8 above)
Comment (danyx23) [reply adminSiteClient/GrapherConfigGridEditor.tsx:390]: Added two helper functions without the boolean flag to make this clearer

File: adminSiteClient/GrapherConfigGridEditor.tsx (hunk 6 above)
Comment (danyx23) [reply adminSiteClient/GrapherConfigGridEditor.tsx:118]: I clarified in a comment why this is needed:  This cell editor component explicitly does nothing. It is used for color scale columns where we don't want to make the cells read only so that copy/paste etc still work but at the same time we don't want the user to be able to edit the cells directly (instead they should use the sidebar editor panel)

File: adminSiteClient/GrapherConfigGridEditorTypesAndUtils.tsx (hunk 4 above)
Comment (danyx23) [reply adminSiteClient/GrapherConfigGridEditorTypesAndUtils.tsx:176]: Deleted it - I thought I might need to have editors be able to edit several properties at once but for now we can make it work with just a single property (as colorScale is a single property that is a JS object but for mapping to columns it's a single entity so the mapping to multiple entities is not needed for now)

File: clientUtils/patchHelper.ts (hunk 2 above)
Comment (danyx23) [reply clientUtils/patchHelper.ts]: Good catch - this code was suggested by github copilot and I was so blown away by it that I accepted it after only checking some common edge cases. I now made a couple of things more explicit (esp array handling) and added tests

File: adminSiteClient/GrapherConfigGridEditor.tsx [hunk 9]
```diff
@@ -91,6 +94,61 @@ import {
...
 // types
 import { SimpleField, Config, ImmutableTree } from "react-awesome-query-builder"
+import codemirror from "codemirror"
+import { UnControlled as CodeMirror } from "react-codemirror2"
+import jsonpointer from "json8-pointer"
+import { EditorColorScaleSection } from "./EditorColorScaleSection.js"
+import { MapChart } from "../grapher/mapCharts/MapChart.js"
+
+function HotColorScaleRenderer(props: Record<string, unknown>) {
+    return <div>Color scale</div>
```
Comment (MarcelGerber) [inline adminSiteClient/GrapherConfigGridEditor.tsx:104]: One last thing: you could use a different color (e.g. grey) here to differentiate this cell from "normal" string cells.

**Review by MarcelGerber — APPROVED** [review]
All's good now. Very nice! 🙌
